### PR TITLE
Use Java 20 as bootjdk for jdknext

### PR DESCRIPTION
**Pull request by keithc-ca**
See https://github.com/eclipse-openj9/openj9/issues/17289.

### CODE REVIEW

**Comment by keithc-ca** [comment]
In draft mode until we ensure build machines are ready for this change.

**Comment by pshipton** [comment]
I don't think this will work. It downloads the JVM from https://api.adoptopenjdk.net/v3/binary/latest/${bootJDKVersion}/ga/${os}/${arch}/jdk/openj9/normal/adoptopenjdk?project=jdk for which there is no openj9 version of 20 (or even 19) yet.

**Comment by pshipton** [comment]
You can modify the download link. I wrote some ramblings about that in https://github.com/eclipse-openj9/openj9/issues/17289#issuecomment-1526100750

**Comment by keithc-ca** [comment]
This new version updates the URL for downloading (new) boot JDKs.
The old URL pattern doesn't seem to work even for versions for which it _could_ redirect to a Semeru build of OpenJ9.

File: buildenv/jenkins/variables/defaults.yml [hunk 1]
```diff
@@ -163,11 +163,11 @@ boot_jdk_default:
       17: '17'
       19: '18'
       20: '18'
-      next: '18'
+      next: '20'
     dir_strip:
       all: '1'
     url:
-      all: "https://api.adoptopenjdk.net/v3/binary/latest/${bootJDKVersion}/ga/${os}/${arch}/jdk/openj9/normal/adoptopenjdk?project=jdk"
+      all: 'https://api.adoptium.net/v3/binary/latest/${bootJDKVersion}/ga/${os}/${arch}/jdk/hotspot/normal/eclipse?project=jdk'
```
Comment (pshipton) [inline buildenv/jenkins/variables/defaults.yml:170]: We should only be using hotspot for `next` and openj9 for the others.

Comment (keithc-ca) [reply buildenv/jenkins/variables/defaults.yml:170]: I tried using "openj9" for 17, I got a 404 error.

Comment (pshipton) [reply buildenv/jenkins/variables/defaults.yml:170]: Perhaps @AdamBrousseau knows, I assume it used to work. It's only needed when we have a new machine that doesn't already have the binary. We could leave versions earlier than `next` as-is for now in order to deliver the jdk21 fix.

Comment (keithc-ca) [reply buildenv/jenkins/variables/defaults.yml:170]: Adam explained that while adoptopenjdk.net redirects to adoptium.net, the redirection doesn't apply to api.adoptopenjdk.net. I'll update this accordingly.

Comment (keithc-ca) [reply buildenv/jenkins/variables/defaults.yml]: Updated.

File: buildenv/jenkins/variables/defaults.yml [hunk 2]
```diff
@@ -163,11 +163,11 @@ boot_jdk_default:
       17: '17'
       19: '18'
       20: '18'
-      next: '18'
+      next: 'https://api.adoptium.net/v3/binary/latest/20/ga/${os}/${arch}/jdk/hotspot/normal/eclipse?project=jdk'
```
Comment (pshipton) [inline buildenv/jenkins/variables/defaults.yml:166]: Does this work? I think you need to put the version here and the url under `url:`

Comment (keithc-ca) [reply buildenv/jenkins/variables/defaults.yml:166]: I think you're right, perhaps this section should look like this?
```
boot_jdk_default:
  boot_jdk:
    location:
      all: '${HOME}/bootjdks'
    version:
      8: '8'
      11: '11'
      17: '17'
      19: '18'
      20: '18'
      next: '20'
    dir_strip:
      all: '1'
    url:
      all: 'https://api.adoptopenjdk.net/v3/binary/latest/${bootJDKVersion}/ga/${os}/${arch}/jdk/openj9/normal/adoptopenjdk?project=jdk'
      next: 'https://api.adoptium.net/v3/binary/latest/${bootJDKVersion}/ga/${os}/${arch}/jdk/hotspot/normal/eclipse?project=jdk'
```
@AdamBrousseau ?

Comment (pshipton) [reply buildenv/jenkins/variables/defaults.yml:166]: Looks good to me, I believe we've done something like that before.

Comment (keithc-ca) [reply buildenv/jenkins/variables/defaults.yml]: Updated.

**Comment by pshipton** [comment]
jenkins compile amac jdknext

**Comment by pshipton** [comment]
Unfortunately it doesn't work, I guess you'll have to specify all the individual versions rather than setting all.

**Comment by pshipton** [comment]
Either than or perhaps the code that uses it could be fixed to recognize the version before using all.

**Comment by keithc-ca** [comment]
> Either than or perhaps the code that uses it could be fixed to recognize the version before using all.

I think the version-specific URL would be fine if this change were loaded early enough; see https://github.com/eclipse-openj9/openj9/blob/master/buildenv/jenkins/common/variables-functions.groovy#L1603.

File: buildenv/jenkins/variables/defaults.yml [hunk 3]
```diff
@@ -163,11 +163,12 @@ boot_jdk_default:
...
       19: '18'
       20: '18'
-      next: '18'
+      next: '20'
     dir_strip:
       all: '1'
     url:
-      all: "https://api.adoptopenjdk.net/v3/binary/latest/${bootJDKVersion}/ga/${os}/${arch}/jdk/openj9/normal/adoptopenjdk?project=jdk"
+      all: 'https://api.adoptopenjdk.net/v3/binary/latest/${bootJDKVersion}/ga/${os}/${arch}/jdk/openj9/normal/adoptopenjdk?project=jdk'
+      next: 'https://api.adoptium.net/v3/binary/latest/${bootJDKVersion}/ga/${os}/${arch}/jdk/hotspot/normal/eclipse?project=jdk'
```
Comment (AdamBrousseau) [inline buildenv/jenkins/variables/defaults.yml:171]: I think `next` needs to be `20`

Comment (keithc-ca) [reply buildenv/jenkins/variables/defaults.yml:171]: Ah, I think I see what you mean; '20': {url} - that makes sense.

Comment (keithc-ca) [reply buildenv/jenkins/variables/defaults.yml]: Updated in 93021cff0d186f56ab97ff9d978dcd57eec7922d.

**Comment by pshipton** [comment]
jenkins compile amac jdknext

**Comment by pshipton** [comment]
Seems we need more changes to handle other builds, I see
```
12:14:49  ++ ls
12:14:49  ++ grep semeru
12:14:49  + sdkFile=
```

**Comment by keithc-ca** [comment]
jenkins compile amac jdknext

File: buildenv/jenkins/common/variables-functions.groovy [hunk 4]
```diff
@@ -1618,7 +1618,7 @@ def download_boot_jdk(bootJDKVersion, bootJDK) {
         sh """
             curl -LJkO ${sdkUrl}
             mkdir -p ${bootJDK}
-            sdkFile=`ls | grep semeru`
+            sdkFile=`ls | grep 'tar\|zip'`
```
[the remark this answers is not in the document]
Comment (keithc-ca) [reply buildenv/jenkins/common/variables-functions.groovy:1621]: Using single quotes keeps the shell from giving the backslash special meaning; I guess it has to get past groovy first.

**Comment by keithc-ca** [comment]
jenkins compile amac jdknext

**Comment by keithc-ca** [comment]
Looks good now:
```
13:24:13  ++ grep 'tar\|zip'
13:24:13  + sdkFile=OpenJDK20U-jdk_aarch64_mac_hotspot_20.0.1_9.tar.gz
13:24:13  + [[ OpenJDK20U-jdk_aarch64_mac_hotspot_20.0.1_9.tar.gz == *zip ]]
13:24:13  + gzip -cd OpenJDK20U-jdk_aarch64_mac_hotspot_20.0.1_9.tar.gz
13:24:13  + tar xof - -C /Users/jenkins/bootjdks/jdk20 --strip=3
13:24:14  + /Users/jenkins/bootjdks/jdk20/bin/java -version
13:24:15  openjdk version "20.0.1" 2023-04-18
13:24:15  OpenJDK Runtime Environment Temurin-20.0.1+9 (build 20.0.1+9)
13:24:15  OpenJDK 64-Bit Server VM Temurin-20.0.1+9 (build 20.0.1+9, mixed mode)
```

**Comment by keithc-ca** [comment]
jenkins compile win jdknext

**Comment by keithc-ca** [comment]
The Windows build failed:
```
13:26:51  + mkdir -p /home/jenkins/bootjdks/jdk20
13:26:51  mkdir: cannot create directory '/home/jenkins/bootjdks/jdk20': Permission denied
```
@AdamBrousseau Why should that be?

**Comment by AdamBrousseau** [comment]
Looks like that dir is owned by Administrator on the new machines.
Fixed them now.

**Comment by AdamBrousseau** [comment]
I restarted your running win build.

**Comment by keithc-ca** [comment]
Is this another _feature_ of new machines?
```
13:41:24  + git clone -b openj9 https://github.com/ibmruntimes/openj9-openjdk-jdk.git . --reference /home/jenkins/openjdk_cache
13:41:24  fatal: destination path '.' already exists and is not an empty directory.
```

**Review by pshipton — APPROVED** [review]

**Comment by keithc-ca** [comment]
@AdamBrousseau Has the problem on Windows build machines that caused the failure noted in https://github.com/eclipse-openj9/openj9/pull/17302#issuecomment-1527891025 been addressed?

**Review by AdamBrousseau — APPROVED** [review]

**Pull request merged**